### PR TITLE
Allow parsing of years greater than 9999 and less than -9999

### PR DESCRIPTION
**Pull request by voughtdq**
Reworked this with the recommendations on the earlier draft pull request.

### CODE REVIEW

File: lib/elixir/lib/calendar/iso.ex [hunk 1]
```diff
@@ -421,6 +436,21 @@ defmodule Calendar.ISO do
...
   end
 
+  defp do_parse_date(<<y, _rest::binary>> = bin, multiplier, :extended) when is_digit_char(y) do
+    case Integer.parse(bin) do
+      {year, <<@ext_date_sep, unquote(match_month_day)>>}
+      when is_big_year(year) and unquote(guard_month_day) ->
+        {month, day} = unquote(read_month_day)
+        parse_formatted_date(year, month, day, multiplier)
+
+      :error ->
```
Comment (voughtdq) [inline lib/elixir/lib/calendar/iso.ex:446]: Wasn't sure about making the last two clauses into a simple `_ -> {:error, :invalid_format}`.

File: lib/elixir/lib/calendar/iso.ex [hunk 2]
```diff
@@ -421,6 +436,21 @@ defmodule Calendar.ISO do
...
+        {:error, :invalid_format}
+
+      {_integer, _rest} ->
+        {:error, :invalid_format}
+    end
```
Comment (voughtdq) [inline lib/elixir/lib/calendar/iso.ex:451]: Wasn't sure about making the last two clauses into a simple `_ -> {:error, :invalid_format}`.

**Comment by voughtdq** [comment]
I know that in the original draft, it was advised to only match on `+` and `-` for big years, but then I'd have to touch the `parse_naive_datetime_guarded/2` clauses as well. I also think there is utility with parsing the digits without a `+` to practice the principle of robustness.

**Comment by josevalim** [comment]
Hi @voughtdq, sorry for the confusion, but I reviewed the PR before your comments on the issue and I agree that, according to ISO8601, you need to agree on a fixed number of digits for the year field, which is not practical for us to do in Elixir. So if you need more digits, then it is best for you to provide your own parsing. Thank you for double checking :)

**Comment by voughtdq** [comment]
Gotcha! It would need to be implemented as an external library.

Sorry to keep bugging you with this, but my bigger concern here is this situation:

```elixir
iex> Date.new!(100_000, 1, 1) |> Date.to_iso8601() |> Date.from_iso8601!()
** (ArgumentError) cannot parse "100000-01-01" as date, reason: :invalid_format
```

The outgoing format is invalid, which I think is surprising. In general, I'd expect the the output of `Date.to_iso8601/1` to be reversible by using `Date.from_iso8601!/1` or `Date.from_iso8601/1` (and vice versa).

**Comment by josevalim** [comment]
We should probably raise if calling to_iso8601 with year outside of the printable range.

**Comment by voughtdq** [comment]
I've been trying to take a stab at this and I'm not sure it's feasible without adding a callback specifically for formatting invalid dates (e.g. `date_to_string_safe` and `datetime_to_string_safe`) to the `Calendar` behaviour.
Raising will break `Inspect` and `String.Chars` without something to differentiate between strict formatting and lax formatting. 

I'll keep thinking about other ways to do it, but how do you feel about adding callbacks for displaying potentially non-formattable dates to `Calendar`?

The less invasive option is something like `to_string_strict` and `to_iso8601_strict` for people who want to opt in.

**Comment by josevalim** [comment]
For now I don’t think those changes are worth it. If someone needs to format those dates, they will need to write their own code.

**Pull request closed without merging**